[Luo Xizi] iP

lxz333 opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
- text based
- easy to learn
- SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/lxz333/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

***And it is FREE!***

Features:
- [x] Managing todos  
- [x] Managing deadlines
- [x] Managing events

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```Java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

lxz333 committed:
current level 1

lxz333 committed:
add the separating line

lxz333 committed:
add 2 classes: execute and task
complete level2

lxz333 committed:
correct a typo

lxz333 committed:
level 3 complete

lxz333 committed:
level 4 completed

lxz333 committed:
level 5 completed
add exception handling

lxz333 committed:
level 6 completed
add delete function

lxz333 committed:
fix an error of todo task

lxz333 committed:
add text-ui-test

lxz333 committed:
add enums

lxz333 committed:
Add some more OOP including Ui, Parser and TaskList

zhengsterz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,83 @@
+        this.tasks.add(task_given);
+    }
+
+    public String msg_of_add(Task task) {
```
Perhaps consider camelCase e.g messageOfAdd?

zhengsterz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,83 @@
+                tasks.size() + " tasks in the list:D";
+    }
+
+    public String print_curr_tasks() {
```
Perhaps consider camelCase e.g PrintCurrTasks?

zhengsterz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,83 @@
+        return res;
+    }
+
+    public String mark_as_done(int index) throws DukeException {
```
Perhaps consider writing some header comments?

JunHao247 replied:
Agree on this point, would help to have comments

zhengsterz commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,18 @@
+public class Event extends Task {
```
Perhaps consider having the suffix -Task added to classes that inherit from Task for clarity? E.g EventTask

zhengsterz commented on src/main/java/Execute.java:
```diff
@@ -0,0 +1,95 @@
+
+public class Execute {
+    public static void checkEmpty(String input, String command) throws DukeException {
+        if (input.length() < command.length() + 1) {
```
I like the use of switch cases its very neat!

zhengsterz reviewed (commented):
Minor naming issues according to the coding standard but the code itself looks good!

lxz333 committed:
Complete level 7 and add more OOP

lxz333 committed:
Add a storage file

lxz333 committed:
Add duke.txt

lxz333 committed:
Merge branch 'branch-Level-7'

JrmCkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,193 @@
+        return "OK, I've marked this task as not done yet:\n" + task.toString();
+    }
+
+    public String delete_msg(int index) throws WrongIndexException {
```
Perhaps consider changing all your methods name into camelCase? So as to align to the coding standard. 🤔

JrmCkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,193 @@
+        storage.loadData(commandList);
+        for (String userInput: commandList) {
+            String[] expressions = userInput.split(" ");
+            String command = expressions[0];
```
Perhaps you would like to parse your command at the start of the method? This could potentially prevent the need of calling multiple "command.equals(...)" at every stage of the if-else statement. 🤔

JrmCkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,193 @@
+            String command = expressions[0];
+
+            try {
+                if (userInput.equals(parser.convertEnum(Command.LIST))) {
```
Considering that this part of code is similar to loadDateFromDisk, perharps you could implement methods instead of writing similar piece of code twice? 🤔

JrmCkh commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,26 @@
+        switch (c) {
+            case LIST:case MARK:case TODO:case EVENT:
+            case DELETE:case UNMARK:case DEADLINE: case BYE:
+                res = c.name().toLowerCase();
```
Perhaps you could split up the cases to improve readability. 🤔 For case clauses, there are no indentation after the switch clause, according to java coding standards.

JrmCkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,193 @@
+                    Event event_task = new Event(userInput);
+                    addTask(event_task);
+                    String tmp = msg_of_add(event_task);
+                } else if (command.equals(parser.convertEnum(Command.DELETE))) {
```
Maybe you would like to use a switch statement instead of a if-else statement? I personally uses a switch clause, which I feel that it makes it neater overall. 💭

JrmCkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,193 @@
+
+    }
+
+    public void loadDataFromDisk() {
```
Perhaps this method could be added under Storage class instead of Duke? 🤔 Especially since loading data can be considered to be an action under storage.

JrmCkh commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,49 @@
+            sc.close();
+        } catch (FileNotFoundException exception) {
+            System.out.println("No data storage file exists.");
+        }
```
I like how you added corner cases such as file not found. I personally would need to include this in my next iteration as well. 👍

JrmCkh reviewed (commented):
Great code overall! 👍

lxz333 committed:
Complete level 8

lxz333 committed:
Merge branch 'branch-level-8'

lxz333 committed:
Divide classes into packages

lxz333 committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

lxz333 committed:
Modify mainClassName in build.gradle

lxz333 committed:
Add JUnit tests

JunHao247 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,198 @@
+        this.taskList.add(task_given);
+    }
+
+    public String msg_of_add(Task task) {
```
camelCase for method names can be used, would be better for both readability and the coding standard

JunHao247 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,198 @@
+                } else {
+                    throw new WeirdInputException();
+                }
+            } catch (WeirdInputException exc)  {
```
Maybe a better Exception name? for better readability

randallnhr replied:
Agreed with this point, maybe you could use something like InvalidInputException?

JunHao247 commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,25 @@
+
+public class Deadline extends Task {
+    protected LocalDate ddl;
+    protected String name;
```
Perhaps a better attribute name to be spelled out for better readability

JunHao247 reviewed (commented):
A few changes to be made for the namings, structure looks good to me

randallnhr commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,27 @@
+        this.isMarked = false;
+    }
+
+    public void mark() {
+        this.isMarked = true;
+    }
```
Good job on abstracting the changing of isMarked boolean to a function! Makes the code much more readable!

lxz333 committed:
Fix some coding standard errors

lxz333 committed:
Complete level 9

lxz333 committed:
Add JavaDoc comments to the code

lxz333 committed:
Merge branch 'branch-A-JavaDoc'

lxz333 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/task/TaskList.java

lxz333 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/task/Parser.java

lxz333 committed:
Add the prototype of GUI

lxz333 committed:
Add fxml

lxz333 committed:
Complete GUI

lxz333 committed:
Modify the way of storing data

lxz333 committed:
Add JavaDoc comment for getResponse() in Duke.java

lxz333 committed:
Merge branch 'branch-Level-10'

lxz333 committed:
Add var-args to support searching for multiple strings

lxz333 committed:
Merge branch 'branch-A-Varargs'

lxz333 committed:
Edit the mainClassName

lxz333 committed:
Add checkStyle to detect coding style violations

lxz333 committed:
Merge branch 'branch-A-CheckStyle'

lxz333 committed:
Add assertions

lxz333 committed:
Edit the method getTasksWanted()

getTasksWanted() is in an arrow-headed style.

This kind of deep nesting makes it hard for readers to
understand the code.

Let's modify the code structure to avoid the arrowhead.

Merging the two "if"s is preferable over having an "if"
nest in another "if".

lxz333 committed:
Improve code quality

There are some coding style issues including bad naming
style, wrong indentations and missing JavaDoc.

These problems make the code messy and break the
general coding standard.

Let's rename the variables and add missing JavaDocs.

lxz333 committed:
Add the method handleUserInput()

loadDataFromDisk() and getResponse(String userInput) have
similar behaviour, which causes code duplication.

Extracting the common behaviour into another method can avoid
duplication and long-winded code.

Let's pull up behaviours common to the 2 methods into a new
parent class named PersonAttribute.

lxz333 committed:
Merge pull request #2 from lxz333/branch-A-CodeQuality

Branch a code quality

lxz333 committed:
Merge branch 'master' of github.com:lxz333/ip

lxz333 committed:
Merge pull request #3 from lxz333/branch-A-Assertions

Add assertions

lxz333 committed:
Merge branch 'master' of github.com:lxz333/ip

lxz333 committed:
Delete run() method

lxz333 committed:
Delete the class Execute

lxz333 committed:
Add more possible exceptions

lxz333 committed:
Handle all possible errors

lxz333 committed:
Improve the GUI

lxz333 committed:
Sort deadlines chronologically and
sort other tasks lexicographically

lxz333 committed:
Fix a bug in the find action

lxz333 committed:
Add hour and minute parsing

lxz333 committed:
Complete README

lxz333 committed:
Add a screenshot for Ui

lxz333 committed:
Use PauseTransition to delay the exiting progress

lxz333 committed:
Add method isCommand()
